ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java

```diff
@@ -112,42 +201,109 @@ public void remove(String blobId) {
     return resolve(blobId, offset, size);
   }
 
-  @Nullable
-  public byte[] resolve(String blobId, int offset, int size) {
+  public @Nullable byte[] resolve(String blobId, int offset, int size) {
     byte[] data = mBlobs.get(blobId);
     if (data == null) {
       return null;
     }
     if (size == -1) {
       size = data.length - offset;
     }
-    if (offset > 0) {
+    if (offset > 0 || size != data.length) {
       data = Arrays.copyOfRange(data, offset, offset + size);
     }
     return data;
   }
 
-  @Nullable
-  public byte[] resolve(ReadableMap blob) {
+  public @Nullable byte[] resolve(ReadableMap blob) {
     return resolve(blob.getString("blobId"), blob.getInt("offset"), blob.getInt("size"));
   }
 
+  private byte[] getBytesFromUri(Uri contentUri) throws IOException {
+    InputStream is = getReactApplicationContext().getContentResolver().openInputStream(contentUri);
+
+    if (is == null) {
+      throw new FileNotFoundException("File not found for " + contentUri);
+    }
+
+    ByteArrayOutputStream byteBuffer = new ByteArrayOutputStream();
+    int bufferSize = 1024;
+    byte[] buffer = new byte[bufferSize];
+    int len;
+    while ((len = is.read(buffer)) != -1) {
+      byteBuffer.write(buffer, 0, len);
+    }
+    return byteBuffer.toByteArray();
+  }
+
+  private String getNameFromUri(Uri contentUri) {
+    if (contentUri.getScheme().equals("file")) {
+      return contentUri.getLastPathSegment();
+    }
+    String[] projection = {MediaStore.MediaColumns.DISPLAY_NAME};
+    Cursor metaCursor = getReactApplicationContext()
+      .getContentResolver()
+      .query(contentUri, projection, null, null, null);
+    if (metaCursor != null) {
+      try {
+        if (metaCursor.moveToFirst()) {
+          return metaCursor.getString(0);
+        }
+      } finally {
+        metaCursor.close();
+      }
+    }
+    return contentUri.getLastPathSegment();
+  }
+
+  private long getLastModifiedFromUri(Uri contentUri) {
+    if (contentUri.getScheme().equals("file")) {
+      return new File(contentUri.toString()).lastModified();
+    }
+    return 0;
+  }
+
+  private String getMimeTypeFromUri(Uri contentUri) {
+    String type = getReactApplicationContext().getContentResolver().getType(contentUri);
```

Comment:
KitKat also has a [`android.provider.DocumentsContract.Document.COLUMN_MIME_TYPE`](https://developer.android.com/reference/android/provider/DocumentsContract.Document.html#COLUMN_MIME_TYPE).

Reply:
Are you sure this is needed, looks like ContentResolver#getType calls the ContentProvider#getType method which is implemented like https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/provider/DocumentsProvider.java#829 for the document one.